### PR TITLE
psh: add kill test and fix reading current target in psh test module

### PR DESCRIPTION
**Pull request by damianloew**
JIRA: PD-77

## Description

- psh: add kill test
- There is also helper executable, which creates 3 while test processes in background (to kill) 
## Motivation and Context

-  test to cover psh kill fixes in phoenix-rtos-kernel (https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/239)
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (ia32-generic - qemu).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/239).
- [ ] I will merge this PR by myself when appropriate.

I will merge this PR after phoenix-rtos-kernel fixes implementation

### CODE REVIEW

File: psh/test-kill.c [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	char *const arg[][9] = {
+		{ "test_proc_0", NULL },
+		{ "test_proc_1", NULL },
+		{ "test_proc_2", NULL }
+	};
+	char *const bsname = basename(argv[0]);
+
+	if (strstr(bsname, "test_proc_")) {
+		while (1)
+			;
```
Comment (agkaminski) [inline psh/test-kill.c:35]: Maybe usleep/sleep here? This will cause 100% CPU usage

Comment (damianloew) [reply psh/test-kill.c]: Thanks, to make it simper, I leave only psh processes testing. I will appreciate, if you have a look second time, to push it.

File: psh/test-kill.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+
+def is_hardware_target(p):
+    cmd = 'ls'
+
+    p.sendline(cmd)
+    p.expect(cmd)
+    idx = p.expect([
+            'syspage',
+            'bin'])
```
Comment (kemonats) [inline psh/test-kill.py:29]: Why not in one line?
```suggestion
    idx = p.expect(['syspage', 'bin'])
```

File: psh/test-kill.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    p.expect(PROMPT)
+
+    if idx == 0:
+        return True
+    elif idx == 1:
+        return False
```
Comment (kemonats) [inline psh/test-kill.py:35]:
```suggestion
    return idx == 0
```

File: psh/test-kill.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        line = p.match.group(1)
+        if not header_seen and line.split() == ps_header:
+            header_seen = True
+        else:
+            try:
+                pid, ppid, pr, state, cpu, wait, time, vmem, thr, task = line.split()
+            except ValueError:
+                assert False, f'wrong ps output: {line}'
+
+            ps_list.append({'pid': pid, 'task': task, 'state': state})
```
Comment (kemonats) [inline psh/test-kill.py:59]: Proposition 1:
```suggestion
        if not header_seen:
            line = line.strip()
            header_seen = line.startswidth('PID')
        else:
            try:
                pid, ppid, pr, state, cpu, wait, time, vmem, thr, task = line.split(maxsplit=9)
            except ValueError:
                assert False, f'wrong ps output: {line}'

            ps_list.append({'pid': pid, 'task': task, 'state': state})
```
The variable `ps_header` is not needed.

Proposition 2:
```suggestion
        try:
            pid, ppid, pr, state, cpu, wait, time, vmem, thr, task = line.split(maxsplit=9)
        except ValueError:
            assert False, f'wrong ps output: {line}'

        if pid.isdigit():
            ps_list.append({'pid': pid, 'task': task, 'state': state})
```
The variables `ph_header` and `header_seen` are not necessary.
`maxsplit=9` is needed when `cmd` is used with arguments.
Maybe `get_process_list(p)` define as a `generator`.

File: psh/test-kill.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    else:
+        cmd = 'sysexec psh'
+
+    for i in range(count):
+        p.sendline(cmd)
+        p.expect(cmd + EOL)
+    p.expect(PROMPT)
+    ps_list = get_process_list(p)
+    # find pid list of sleep psh processes
+    for proc in ps_list:
```
Comment (kemonats) [inline psh/test-kill.py:77]:
```suggestion
    # find pid list of sleep psh processes
    for proc in get_process_list(p):
```

File: psh/test-kill.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    psh_pid_count = len(psh_pid_list)
+    assert psh_pid_count == count, f'Created {psh_pid_count} psh processes, instead of {count}'
+
+    return psh_pid_list
+
+
+def assert_kill_procs(p, pid_list, pid_count):
+    for i in range(pid_count):
+        cmd = f'kill {pid_list[i]}'
```
Comment (kemonats) [inline psh/test-kill.py:90]:
```suggestion
    for pid in pid_list):
        cmd = f'kill {pid}'
```

File: psh/test-kill.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        p.sendline(cmd)
+        p.expect(cmd + EOL)
+    sleep(0.5)
+    ps_list = get_process_list(p)
+
+    for nr in range(pid_count):
+        assert {'pid': pid_list[nr], 'task': f'test_proc_{nr}'} not in ps_list, \
+            f'The process with id: {pid_list[nr]} is still in the processes list, even though it has been killed'
```
Comment (kemonats) [inline psh/test-kill.py:98]:
```suggestion
    for proc in get_process_list(p):
        assert not proc['pid'] in pid_list, \
            f'The process with id: {proc["pid"]} is still in the processes list, even though it has been killed'
```

File: psh/test-kill.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    pid = 90
+    used_pid_list = []
+    ps_list = get_process_list(p)
+
+    for proc in ps_list:
+        used_pid_list.append(proc.get('pid'))
+
+    # if process id is used, find next unused id
+    while f'{pid}' in used_pid_list:
+        pid = pid + 1
```
Comment (kemonats) [inline psh/test-kill.py:111]:
```suggestion
    max_pid = max([int(d["pid"]) for d in get_process_list(p)])
    if max_pd >= pid:
        pid += 1
```

File: psh/test-kill.py (hunk 4 above)
Comment (damianloew) [reply psh/test-kill.py:43]: You are right, thanks

File: psh/test-kill.py (hunk 8 above)
Comment (damianloew) [reply psh/test-kill.py]: Thanks for Your suggestions

File: psh/test-kill.c (hunk 1 above)
Comment (damianloew) [reply psh/test-kill.c]: Finally I removed additional c executable

File: psh/test-kill.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    used_pid_list = [proc["pid"] for proc in get_process_list(p)]
+
+    # if process id is used, find next unused id
+    while str(pid) in used_pid_list:
+        pid = pid + 1
+
+    msg = 'Wrong kill help message'
+    psh.assert_cmd(p, 'kill', 'usage: kill <pid>', msg)
+    msg = 'Wrong kill error message when passing string in argument'
+    psh.assert_cmd(p, 'kill not_number', 'kill: could not parse process id: not_number', msg)
```
Comment (xvuko) [inline psh/test-kill.py:81]: I do not like checking exact error message output. Ideally should be done by checking return value but i assume this is currently not possible on psh.

Comment (damianloew) [reply psh/test-kill.py:81]: Good suggestion, there is some idea of creating additional psh command, which would return exit code. Once this is done, I will use it and possibly change my current tests.

**Comment by damianloew** [comment]
@xvuko thanks for review, I've applied changes in the last force push.

File: psh/test-kill.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+def assert_kill_procs(p, pid_list, pid_count):
+    for pid in pid_list:
+        msg = f'Wrong output when killing process with the following pid: {pid}'
+        psh.assert_cmd(p, f'kill {pid}', '', msg)
+    sleep(0.5)
+
+    for proc in get_process_list(p):
+        assert not proc['pid'] in pid_list, \
+            f'The process with id: {proc["pid"]} is still in the processes list, even though it has been killed'
```
[the remark this answers is not in the document]
Comment (damianloew) [reply psh/test-kill.py]: set(processes to kill) - set(current processes) will be dead processes, so I've changed it a little bit.

**Comment by damianloew** [comment]
@xvuko thanks for review, I've applied changes in the last force push.

**Pull request merged**